components/omega/doc/devGuide/Config.md

```diff
@@ -0,0 +1,118 @@
+the configuration is read in and owned by the master MPI task and the
+retrieval from the full configuration into module variables involves a
+broadcast to the rest of the tasks. In addition, this paradigm will allow
+the deletion of the configuration at the end of initialization to free memory.
+
+The configuration is built around the yaml-cpp library. Each configuration
+contains a ``YAML::Node`` and a name for the configuration.
+Most of the functions below are wrappers around yaml-cpp templated functions.
+Rather than using templated forms in omega, we alias the function interfaces
+so that the syntax is cleaner.
+
+Within Omega, the full configuration file is read just after the MachEnv
+initialization using:
+```c++
+OMEGA::Config::readAll("omega.yml");
+```
+The full Omega configuration is stored in a static variable for later
+retrievals.
+
+Each module in Omega will extract its own configuration variables by
+first retrieving the stored Omega configuration, then retrieving the module
+configuration (multiple times if the module is nested under a parent)
+and finally retrieving the variable by name. The sequence might look like
+this for the example shown in the User Guide:
+```c++
+// Get pointer to full configuration
+OMEGA::Config *OmegaConfig = OMEGA::Config::getOmegaConfig;
+
+OMEGA::Config HmixConfig; // creates an empty config for Hmix
+
+OmegaConfig->get("Hmix",HmixConfig); // extracts all Hmix config vars from omega
+
+// Get individual variables
+Err = HmixConfig.get("HmixScaleWithMesh", HmixScaleWithMesh);
+Err = HmixConfig.get("MaxMeshDensity"   , MaxMeshDensity);
+Err = HmixConfig.get("HmixUseRefWidth"  , HmixUseRefWidth);
+Err = HmixConfig.get("HmixRefWidth"     , HmixRefWidth);
+```
+If the module is deeper in the hierarchy (eg an HmixDel2 module
+under the Hmix module), you would need an additional call to extract
+the subconfiguration, like:
+```c++
+OMEGA::Config HmixDel2Config; // creates and empty config for Hmix Del2
+Err = HmixConfig.get(HmixDel2Config);
+```
+If the variable cannot be found in the input configuration, a non-zero
+error code is returned.  Existence can also be checked with the bool
+functions existsGroup or existsVar, where the variable name is provided.
+The developer will need to decide the action to take based on that
+condition. For example, the code could return a critical error and terminate
+if it is required. In some cases, it may be sufficient to issue a warning
+and set the value of the variable to an internal default.
+In this latter case, it is a good idea to add the variable and its value to
+the configuration using the add function. For example:
+```c++
+Err = HmixConfig.add("HmixNewVar", value);
+```
+Then it will be included in any future writing of the configuration in
+order to save provenance. Similarly, an existing value can be changed using
+the ``set`` function that has an analogous form. There is a also remove
```

Comment:
Would there be any way to tell when in the simulation a config option was changed? Maybe a flag for startup vs. runtime?

Reply:
There is a capability for writing the config and I was assuming that we might write the full (potentially modified) config at the end of the init routine for provenance purposes. I was also assuming that this would still all happen during the init phase, but the module does not preclude the possibility of run-time changes as long as the config hasn't been deleted.  In that case, we could decide whether to re-write the config or just output a message to the log.

Reply:
Fixed this typo too

Reply:
Sounds good. I'm thinking about safeguards for a situation where a developer doesn't intend to change a config option during runtime but they put a config-set in the wrong place (probably hard to do accidentally) or where a user doesn't realize that a config option has been changed during runtime. Would it make sense to have a check at the end of runtime that the config options haven't been changed from their values at the end of init and output a warning to the log? Not sure whether that's overkill.

It might also be helpful to a user to have any easy way to check whether values have been changed during init. A warning message would potentially be more helpful than having to do the separate step of diff-ing yaml files, but maybe you have a better idea?